### ISSUE TITLE
Tenant slug doesn't work

### ISSUE THREAD

**Issue by JupiterThreads** [Corvia/django-tenant-users#111]
If I try to create a tenant with the slug of 'dunder-mifflin', I get a validation error from django tenants:
`ValidationError("Invalid string used for the schema name.")`

It looks like its because django-tenant-users is creating a schema name on line 49 in the provision_tenant function from the tenant slug, which therefore means the passed in slug must contain valid postgres schema characters only. But of course valid django slugs can contain hyphens, and django tenants does allow you use to them in a domain.

I propose some regex that strips out all hyphens and underscores from the slug and use that as the schema name prefix. Or just replace all hyphens with underscores.

**Comment by Dresdn** [comment]
The `ValidationError` is coming from `django-tenants`, which is the proper way to handle it. Instead, I would advocate validating the input within your application as changing the slug on the fly could lead to unintended side effects.

Curious your thoughts and how you're calling `provision_tenant()`

**Comment by JupiterThreads** [comment]
I don't agree. I think that error is highlighting an inherent issue in django-tenant-users. One of the provision_tenant's arguments is a tenant_slug so the caller would rightly expect to pass in a valid slug according to django's docs:

https://docs.djangoproject.com/en/3.2/ref/models/fields/
> A slug is a short label for something, containing only letters, numbers, underscores or hyphens. They’re generally used in URLs.

However, on line 49 we can see the tenant_slug is being used to construct the schem_name and there's even a comment saying that it must contain a valid postgres schema name: 

```
time_string = str(int(time.time()))
    # Must be valid postgres schema characters see:
    # https://www.postgresql.org/docs/9.2/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
    # We generate unique schema names each time so we can keep tenants around
    # without taking up url/schema namespace.
    schema_name = '{0}_{1}'.format(tenant_slug, time_string)
```

I don't think its reasonable for a caller to know that a tenant_slug they're passing in should also be a valid schema_name and that a valid tenant slug is no longer valid. 

The problem is due to this function taking on the responsibility of creating the schema_name instead of having it is a separate argument, which is why I think mutating the tenant_slug before its used in the schema_name would have the least impact. If it was not valid, it would still throw the same error as it does now. The tenant_slug would not change. For example,

```
schema_name_prefix = tenant_slug.replace('-', '_')
schema_name = f'{schane_name_prefix}_{time_string}'

```
Without this change, I then have to do something like this:

```
tenant_slug = 'dunder-mifflin'
fake_slug = tenant_slug.replace('-', '_')
provision_tenant(name, fake_slug, owner.email)
domain = get_object_or_404(Domain, domain=fake_slug)
domain.domain = tenant_slug # I shouldn't be able to do this if provision_tenant doesn't let me
domain.save(update_fields=['domain'])
return domain.tenant
```

**Comment by Dresdn** [comment]
I understand all your reasons, and completely agree with them. However, I still believe that the logic should be handled by your application in a way that works for you. Changing data could lead to unmet expectations and behavior, creating more problems than it solves. For example, if you're using `django_tenants` with the `TENANT_SUBFOLDER_PREFIX`, thinking my tenant path is `http://domain.com/tenant-name` when it is automagically "corrected" to `http://domain.com/tenant_name` isn't appropriate.

As you didn't answer my question on how you're creating the tenant, I'm going to assume here. Instead, I would recommend either handling the `ValidationError` with a `try...except` or use `django_tenants.postgresql_backend.base.is_valid_schema_name()` and then run your regex before calling `provision_tenant()`

Lastly, if you'd really like this changed, I would suggest opening an issue at the [django-tenants page](https://github.com/django-tenants/django-tenants) as they own the exception throwing.

**Comment by productdev-catalyst** [comment]
So I'm confused how do you think my suggestion would change the tenant path exactly? 

If I pass in 'tenant-name' as a tenant_slug, tenant_domain will equal 'tenant-name' for TENANT_SUBFOLDER_PREFIX and for non-subfolders it will still equal  '{0}.{1}'.format(tenant_slug, settings.TENANT_USERS_DOMAIN), which gets saved to the domain model only on line 61.

So you will have a path `http://domain.com/tenant-name` and there will be no change to the domain, and you will have catered for the valid hyphen. The schema name is never being used as part of the domain, and even if it was it would look something like this: http://domain.com/tenant_name_1636127387

The ONLY thing that will change will be the schema_name in the database that gets saved on the TenantModel, and you already constructing that in the provision_tenant function.

I don't mind opening a ticket with django-tenants, but what would I say when I can clearly see the error is starting from line 68 in provision_tenant function:
```
\.virtualenvs\coeus_prime_backend-d8nHQZ8s\lib\site-packages\tenant_users\tenants\tasks.py", line 68, in provision_tenant
    tenant = TenantModel.objects.create(
```

Sorry I didn't answer your question, but I'm not calling provision_tenant in any special way. Its just: 

`provision_tenant(name, tenant_slug, owner.email)
`

Out of curiosity, how do you create a tenant with two words, e.g 'evil corp'. Have you ever tried:
`fqdn = provision_tenant("EvilCorp", "evil-corp", "admin@evilcorp.com").
`
I think its going to be a very common issue. I've only started using this app in the past month and I've had to create multiple tenants with more than one word in their name. Having to use the underscore and introducing a hack/workaround just makes no sense to me when there's an easily solution.

**Comment by JupiterThreads** [comment]
So I'm confused how do you think my suggestion would change the tenant path exactly? 

If I pass in `tenant-name` as a tenant_slug, tenant_domain will equal `tenant-name` for TENANT_SUBFOLDER_PREFIX and for non-subfolders it will still equal  `{0}.{1}'.format(tenant_slug, settings.TENANT_USERS_DOMAIN)`, which gets saved to the domain model only on line 61.

So you will have a path `http://domain.com/tenant-name` and there will be no change to the domain, and you will have catered for the valid hyphen. The schema name is never being used as part of the domain, and even if it was it would look something like this: `http://domain.com/tenant_name_1636127387`

The ONLY thing that will change will be the schema_name in the database that gets saved on the TenantModel, and you are already constructing that in the provision_tenant function.

I don't mind opening a ticket with django-tenants, but what would I say when I can clearly see the error is starting from line 68 in provision_tenant function:
```
\.virtualenvs\coeus_prime_backend-d8nHQZ8s\lib\site-packages\tenant_users\tenants\tasks.py", line 68, in provision_tenant
    tenant = TenantModel.objects.create(
```

Sorry I didn't answer your question, but I'm not calling provision_tenant in any special way. Its just: 

`provision_tenant(name, tenant_slug, owner.email)
`

Out of curiosity, how do you create a tenant with two words, e.g 'evil corp'. Have you ever tried:
`fqdn = provision_tenant("EvilCorp", "evil-corp", "admin@evilcorp.com").
`
I think its going to be a very common issue. I've only started using this app in the past month and I've had to create multiple tenants with more than one word in their name. Having to use the underscore and introducing a hack/workaround just makes no sense to me when there's an easily solution.

**Comment by Dresdn** [comment]
I need to eat crow on this one, as I completely misunderstood your original ask. I've been replying under the assumption you meant to change the slug itself, and therefore the schema. My apologies for the confusion!

To get back to it, this is actually interesting. Looking at the [PostgreSQL Schema Create docs](https://www.postgresql.org/docs/13/sql-createschema.html), there is no reason that a slug of "dunder-mifflin" creating a schema of `dunder-mifflin_01234` should throw an error. To use non `[a-z][0-9]_` characters are valid, as long as you quote the schema name, which is [exactly what django-tenants does](https://github.com/django-tenants/django-tenants/blob/master/django_tenants/models.py#L197).

I went back to PostgreSQL 8.x docs and the same rules apply, and the code was part of the original fork in 2015, so I'm wondering if this is just old logic that's no longer applicable.

Long story short, I see your point and agree it should be addressed. How about we do a dual path approach? We can do an RFC 1123 check on `slug` when used as a hostname, or RFC 3986 3.3 when `TENANT_SUBFOLDER_PREFIX` is defined (hostname valid chars are slightly different than URL path chars).

In parallel, I do believe `django-tenants` should allow for dashes and spaces in the schema name, so I don't mind opening up an issue there (unless you wanna do it?) to help improve that project.

To summarize work that can be done here, there's a couple todos:

[ ] Add RFC validation for `tenant_slug`
[ ] Convert any chars that don't match [`django_tenants.postgresql_backend.base.is_valid_schema_name()`](https://github.com/django-tenants/django-tenants/blob/master/django_tenants/postgresql_backend/base.py#L28) to an underscore when [creating `schema_name`](https://github.com/Corvia/django-tenant-users/blob/master/tenant_users/tenants/tasks.py#L44)

Sound like a plan?

**Comment by JupiterThreads** [comment]
ha, no worries. It's usually me with the confusion.

Ah, I've just tried it out and you are right, non `[a-z][0-9_`  characters can be used for the schema name. I'm surprised no one has picked up on it. 

That sounds good. I have raised a ticket with django-tenants, https://github.com/django-tenants/django-tenants/issues/688. Feel free to add any more to it in the comments. Let me know if you want me to do anything else. I like this app and I'm keen on seeing it do it well. 

Thanks a lot.
